File: tests/docker-compat/js-client/tests/advanced.js
const { register, runTest, key, assertEqual, assertTrue, assertEqualInt, assertEqualFloat } = require('./framework');

register('advanced_types', async (redis, prefix) => {
  const results = [];
  const cat = 'advanced_types';

  // --- Bitmap ---

  results.push(await runTest('SETBIT_GETBIT', cat, async () => {
    const k = key(prefix, 'bm:sg');
    const old = await redis.setbit(k, 7, 1);
    assertEqualInt(0, old);
    const val = await redis.getbit(k, 7);
    assertEqualInt(1, val);
    const val0 = await redis.getbit(k, 0);
    assertEqualInt(0, val0);
    await redis.del(k);
  }));

  results.push(await runTest('BITCOUNT', cat, async () => {
    const k = key(prefix, 'bm:count');
    await redis.set(k, 'foobar');
    const n = await redis.bitcount(k);
    assertEqualInt(26, n);
    await redis.del(k);
  }));

  results.push(await runTest('BITOP', cat, async () => {
    const k1 = key(prefix, 'bm:op1');
    const k2 = key(prefix, 'bm:op2');
    const dest = key(prefix, 'bm:opdest');
    await redis.set(k1, 'abc');
    await redis.set(k2, 'def');
    const n = await redis.bitop('AND', dest, k1, k2);
    assertEqualInt(3, n);
    await redis.del(k1, k2, dest);
  }));

  results.push(await runTest('BITPOS', cat, async () => {
    const k = key(prefix, 'bm:pos');
    await redis.setbit(k, 0, 0);
    await redis.setbit(k, 1, 0);
    await redis.setbit(k, 2, 1);
    const pos = await redis.bitpos(k, 1);
    assertEqualInt(2, pos);
    await redis.del(k);
  }));

  // --- HyperLogLog ---

  results.push(await runTest('PFADD_PFCOUNT', cat, async () => {
    const k = key(prefix, 'hll:basic');
    const n = await redis.pfadd(k, 'a', 'b', 'c', 'd');
    assertTrue(n > 0, 'PFADD should return 1 for new elements');
    const count = await redis.pfcount(k);
    assertTrue(count >= 3 && count <= 5, 'PFCOUNT should be ~4');
    await redis.del(k);
  }));

  results.push(await runTest('PFMERGE', cat, async () => {
    const k1 = key(prefix, 'hll:m1');
    const k2 = key(prefix, 'hll:m2');
    const dest = key(prefix, 'hll:merged');
    await redis.pfadd(k1, 'a', 'b', 'c');
    await redis.pfadd(k2, 'c', 'd', 'e');
    await redis.pfmerge(dest, k1, k2);
    const count = await redis.pfcount(dest);
    assertTrue(count >= 4 && count <= 6, 'merged count should be ~5');
    await redis.del(k1, k2, dest);
  }));

  // --- Geo ---

  results.push(await runTest('GEOADD_GEOPOS', cat, async () => {
    const k = key(prefix, 'geo:cities');
    const n = await redis.geoadd(k, 2.3522, 48.8566, 'Paris', -0.1278, 51.5074, 'London', 13.4050, 52.5200, 'Berlin');
    assertEqualInt(3, n);
    const positions = await redis.geopos(k, 'Paris', 'London');
    assertEqualInt(2, positions.length);
    assertTrue(positions[0] !== null, 'Paris position should not be null');
    await redis.del(k);
  }));

  results.push(await runTest('GEODIST', cat, async () => {
    const k = key(prefix, 'geo:dist');
    await redis.geoadd(k, 2.3522, 48.8566, 'Paris', -0.1278, 51.5074, 'London');
    const dist = await redis.geodist(k, 'Paris', 'London', 'km');
    const d = parseFloat(dist);
    assertTrue(d > 300 && d < 400, 'Paris-London distance should be ~340km');
    await redis.del(k);
  }));

  results.push(await runTest('GEOSEARCH', cat, async () => {
    const k = key(prefix, 'geo:search');
    await redis.geoadd(k, 2.3522, 48.8566, 'Paris', -0.1278, 51.5074, 'London', 13.4050, 52.5200, 'Berlin', -3.7038, 40.4168, 'Madrid');
    const members = await redis.call('GEOSEARCH', k, 'FROMLONLAT', 2.3522, 48.8566, 'BYRADIUS', 500, 'km', 'ASC');
    assertTrue(members.length >= 1, 'should find at least Paris');
    assertEqual('Paris', members[0]);
    await redis.del(k);
  }));

  results.push(await runTest('GEOHASH', cat, async () => {
    const k = key(prefix, 'geo:hash');
    await redis.geoadd(k, 2.3522, 48.8566, 'Paris');
    const hashes = await redis.geohash(k, 'Paris');
    assertEqualInt(1, hashes.length);
    assertTrue(hashes[0].length > 0, 'geohash should not be empty');
    await redis.del(k);
  }));

  results.push(await runTest('GEOHASH_accuracy', cat, async () => {
    const k = key(prefix, 'geo:hash_acc');
    await redis.geoadd(k, 2.3522, 48.8566, 'Paris');
    const hashes = await redis.geohash(k, 'Paris');
    assertTrue(hashes[0].startsWith('u0'), "Paris geohash should start with 'u0'");
    await redis.del(k);
  }));

  return results;
});

File: tests/docker-compat/python-client/tests/advanced.py
"""Category 13: Advanced Data Types tests (Bitmap, HyperLogLog, Geo)."""

from .framework import (
    register, run_test, key, assert_equal, assert_true,
    assert_equal_int, assert_equal_float,
)


def test_advanced(r, prefix):
    results = []
    cat = "advanced_types"

    # --- Bitmap ---

    def t_setbit_getbit():
        k = key(prefix, "bm:sg")
        try:
            old = r.setbit(k, 7, 1)
            assert_equal_int(0, old)
            val = r.getbit(k, 7)
            assert_equal_int(1, val)
            val = r.getbit(k, 0)
            assert_equal_int(0, val)
        finally:
            r.delete(k)
    results.append(run_test("SETBIT_GETBIT", cat, t_setbit_getbit))

    def t_bitcount():
        k = key(prefix, "bm:count")
        try:
            r.set(k, "foobar")
            n = r.bitcount(k)
            assert_equal_int(26, n)
        finally:
            r.delete(k)
    results.append(run_test("BITCOUNT", cat, t_bitcount))

    def t_bitop():
        k1 = key(prefix, "bm:op1")
        k2 = key(prefix, "bm:op2")
        dest = key(prefix, "bm:opdest")
        try:
            r.set(k1, "abc")
            r.set(k2, "def")
            n = r.bitop("AND", dest, k1, k2)
            assert_equal_int(3, n)
        finally:
            r.delete(k1, k2, dest)
    results.append(run_test("BITOP", cat, t_bitop))

    def t_bitpos():
        k = key(prefix, "bm:pos")
        try:
            r.setbit(k, 0, 0)
            r.setbit(k, 1, 0)
            r.setbit(k, 2, 1)
            pos = r.bitpos(k, 1)
            assert_equal_int(2, pos)
        finally:
            r.delete(k)
    results.append(run_test("BITPOS", cat, t_bitpos))

    # --- HyperLogLog ---

    def t_pfadd_pfcount():
        k = key(prefix, "hll:basic")
        try:
            n = r.pfadd(k, "a", "b", "c", "d")
            assert_true(n > 0, "PFADD should return 1 for new elements")
            count = r.pfcount(k)
            assert_true(3 <= count <= 5, "PFCOUNT should be ~4")
        finally:
            r.delete(k)
    results.append(run_test("PFADD_PFCOUNT", cat, t_pfadd_pfcount))

    def t_pfmerge():
        k1 = key(prefix, "hll:m1")
        k2 = key(prefix, "hll:m2")
        dest = key(prefix, "hll:merged")
        try:
            r.pfadd(k1, "a", "b", "c")
            r.pfadd(k2, "c", "d", "e")
            r.pfmerge(dest, k1, k2)
            count = r.pfcount(dest)
            assert_true(4 <= count <= 6, "merged count should be ~5")
        finally:
            r.delete(k1, k2, dest)
    results.append(run_test("PFMERGE", cat, t_pfmerge))

    # --- Geo ---

    def t_geoadd_geopos():
        k = key(prefix, "geo:cities")
        try:
            n = r.geoadd(k, [2.3522, 48.8566, "Paris", -.1278, 51.5074, "London", 13.4050, 52.5200, "Berlin"])
            assert_equal_int(3, n)
            positions = r.geopos(k, "Paris", "London")
            assert_equal(2, len(positions))
            assert_true(positions[0] is not None, "Paris position should not be None")
        finally:
            r.delete(k)
    results.append(run_test("GEOADD_GEOPOS", cat, t_geoadd_geopos))

    def t_geodist():
        k = key(prefix, "geo:dist")
        try:
            r.geoadd(k, [2.3522, 48.8566, "Paris", -.1278, 51.5074, "London"])
            dist = r.geodist(k, "Paris", "London", unit="km")
            assert_true(300 < float(dist) < 400, "Paris-London distance should be ~340km")
        finally:
            r.delete(k)
    results.append(run_test("GEODIST", cat, t_geodist))

    def t_geosearch():
        k = key(prefix, "geo:search")
        try:
            r.geoadd(k, [
                2.3522, 48.8566, "Paris",
                -.1278, 51.5074, "London",
                13.4050, 52.5200, "Berlin",
                -3.7038, 40.4168, "Madrid",
            ])
            results_geo = r.geosearch(k, longitude=2.3522, latitude=48.8566, radius=500, unit="km", sort="ASC")
            assert_true(len(results_geo) >= 1, "should find at least Paris")
            assert_equal("Paris", results_geo[0])
        finally:
            r.delete(k)
    results.append(run_test("GEOSEARCH", cat, t_geosearch))

    def t_geohash():
        k = key(prefix, "geo:hash")
        try:
            r.geoadd(k, [2.3522, 48.8566, "Paris"])
            hashes = r.geohash(k, "Paris")
            assert_equal(1, len(hashes))
            assert_true(len(hashes[0]) > 0, "geohash should not be empty")
        finally:
            r.delete(k)
    results.append(run_test("GEOHASH", cat, t_geohash))

    def t_geohash_accuracy():
        k = key(prefix, "geo:hash_acc")
        try:
            r.geoadd(k, [2.3522, 48.8566, "Paris"])
            hashes = r.geohash(k, "Paris")
            assert_true(hashes[0].startswith("u0"), "Paris geohash should start with 'u0'")
        finally:
            r.delete(k)
    results.append(run_test("GEOHASH_accuracy", cat, t_geohash_accuracy))

    return results


register("advanced_types", test_advanced)
